crates/neon/src/types_impl/buffer/types.rs

```diff
@@ -0,0 +1,783 @@
+        //         buffer handle matches the lifetime of the typed array handle.
+        let info = unsafe { sys::typedarray::info(env.to_raw(), local) };
+
+        Self(JsTypedArrayInner {
+            local,
+            buffer: info.buf,
+            _type: PhantomData,
+        })
+    }
+}
+
+impl<T: Binary> TypedArray for JsTypedArray<T> {
+    type Item = T;
+
+    fn as_slice<'cx, 'a, C>(&self, cx: &'a C) -> &'a [Self::Item]
+    where
+        C: Context<'cx>,
+    {
+        unsafe {
+            let env = cx.env().to_raw();
+            let value = self.to_raw();
+            let info = sys::typedarray::info(env, value);
+
+            slice::from_raw_parts(info.data.cast(), info.length)
+        }
+    }
+
+    fn as_mut_slice<'cx, 'a, C>(&mut self, cx: &'a mut C) -> &'a mut [Self::Item]
+    where
+        C: Context<'cx>,
+    {
+        unsafe {
+            let env = cx.env().to_raw();
+            let value = self.to_raw();
+            let info = sys::typedarray::info(env, value);
+
+            slice::from_raw_parts_mut(info.data.cast(), info.length)
+        }
+    }
+
+    fn try_borrow<'cx, 'b, C>(
+        &self,
+        lock: &'b Lock<'b, C>,
+    ) -> Result<Ref<'b, Self::Item>, BorrowError>
+    where
+        C: Context<'cx>,
+    {
+        unsafe {
+            let env = lock.cx.env().to_raw();
+            let value = self.to_raw();
+            let info = sys::typedarray::info(env, value);
+
+            // The borrowed data must be guarded by `Ledger` before returning
+            Ledger::try_borrow(
+                &lock.ledger,
+                slice::from_raw_parts(info.data.cast(), info.length),
+            )
+        }
+    }
+
+    fn try_borrow_mut<'cx, 'a, C>(
+        &mut self,
+        lock: &'a Lock<'a, C>,
+    ) -> Result<RefMut<'a, Self::Item>, BorrowError>
+    where
+        C: Context<'cx>,
+    {
+        unsafe {
+            let env = lock.cx.env().to_raw();
+            let value = self.to_raw();
+            let info = sys::typedarray::info(env, value);
+
+            // The borrowed data must be guarded by `Ledger` before returning
+            Ledger::try_borrow_mut(
+                &lock.ledger,
+                slice::from_raw_parts_mut(info.data.cast(), info.length),
+            )
+        }
+    }
+
+    fn size<'cx, C: Context<'cx>>(&self, cx: &mut C) -> usize {
+        self.len(cx) * std::mem::size_of::<Self::Item>()
+    }
+}
+
+impl<T: Binary> JsTypedArray<T> {
+    /// Constructs a typed array that views `buffer`.
+    ///
+    /// The resulting typed array has `(buffer.size() / size_of::<T>())` elements.
+    pub fn from_buffer<'cx, 'b: 'cx, C>(
+        cx: &mut C,
+        buffer: Handle<'b, JsArrayBuffer>,
+    ) -> JsResult<'cx, Self>
+    where
+        C: Context<'cx>,
+    {
+        let size = buffer.size(cx);
+        let elt_size = std::mem::size_of::<T>();
+        let len = size / elt_size;
+
+        if (len * elt_size) != size {
+            panic!(
```

Comment:
I'm open to anything, and a `RangeError` is certainly consistent with what the JS constructors do (e.g. `new Uint32Array(new ArrayBuffer(7))` throws a `RangeError`). I think an error other than `JsResult` would make the API too painful, so that leaves either panicking or throwing a `RangeError`.

I guess my only question is whether it's weird to throw a JavaScript exception from something that seems more like a pure Rust operation. But I think since it's trying to allocate a new JS object, it's reasonable for that to throw a JS exception. So why don't we go with throwing a `RangeError`.